### ISSUE TITLE
Architecture details of cheat.sh

### ISSUE THREAD

**Issue by rahulunair** [chubin/cheat.sh#231]
Is there any document describing a code walkthrough and details of the architecture choices of cheat.sh?

**Comment by chubin** [comment]
Not really. You are asking about something particular, or a high-level overview would be fine?

**Comment by abitrolly** [comment]
A particular question I have is about missing `/bin/upstream`. Without it, local server tries to contact https://cheat.sh which is giving 500 errors that are impossible to debug, For example the query https://cheat.sh/g++%20/file

Because `bin/upstream` is absent, the lookup adapter is executed always with no fallback https://github.com/chubin/cheat.sh/blob/0ff765a0a3905fbb071a47b83be4d0e575ad69a0/lib/adapter/question.py#L62

**Comment by xinyii** [comment]
I want to know the details of how to get the answers in StackOverflow. This part seems to be in `/bin/upstream`
